File: 프로그래머스/5/49190. 방의 개수/방의 개수.js
function solution(arrows) {
    const direction = [
        [0, 2],
        [2, 2], 
        [2, 0],
        [2, -2],
        [0, -2],
        [-2, -2],
        [-2, 0], 
        [-2, 2],
    ];

    const visitedNodes = new Set();
    const visitedEdges = new Set();

    let x = 0, y = 0;
    visitedNodes.add(`${x},${y}`);
    let roomCount = 0;
 
    for (let arrow of arrows) {
        const dx = direction[arrow][0];
        const dy = direction[arrow][1];

        for (let i = 0; i < 2; i++) {
            const nx = x + dx / 2;
            const ny = y + dy / 2;

            const currentNode = `${x},${y}`;
            const nextNode = `${nx},${ny}`;
            const edge = `${currentNode}->${nextNode}`;
            const reverseEdge = `${nextNode}->${currentNode}`;

            if (!visitedEdges.has(edge)) {
                visitedEdges.add(edge);
                visitedEdges.add(reverseEdge);
                if (visitedNodes.has(nextNode)) {
                    roomCount++;
                }
            }

            visitedNodes.add(nextNode);
            x = nx;
            y = ny;
        }
    }

    return roomCount;
}

File: 프로그래머스/5/49190. 방의 개수/방의 개수.kt
class Solution {
    fun solution(arrows: IntArray): Int {
    val dx = intArrayOf(0, 1, 1, 1, 0, -1, -1, -1)
    val dy = intArrayOf(1, 1, 0, -1, -1, -1, 0, 1)
    var x = 0
    var y = 0
    val visitedNodes = mutableSetOf<Pair<Int, Int>>()
    val visitedEdges = mutableSetOf<Pair<Pair<Int, Int>, Pair<Int, Int>>>()
    visitedNodes.add(Pair(x, y))
    var answer = 0
    for (arrow in arrows) {
        repeat(2) {
            val nx = x + dx[arrow]
            val ny = y + dy[arrow]
            val currentNode = Pair(x, y)
            val nextNode = Pair(nx, ny)
            val edge = Pair(currentNode, nextNode)
            val reverseEdge = Pair(nextNode, currentNode)
            if (!visitedEdges.contains(edge)) {
                if (visitedNodes.contains(nextNode)) {
                    answer++
                }
                visitedEdges.add(edge)
                visitedEdges.add(reverseEdge)
            }
            visitedNodes.add(nextNode)
            x = nx
            y = ny
        }
    }
    return answer
}

}